web/modules/custom/dpl_related_content/src/Services/RelatedContent.php

```diff
@@ -0,0 +1,371 @@
+   *   A possible entity UUID, that will not get included in results.
+   * @param array<int> $tags
+   *   Tag term IDs, to look for.
+   * @param array<int> $categories
+   *   Category term IDs, to look for.
+   * @param array<int> $branches
+   *   Branch term IDs, to look for.
+   *
+   * @return array<int|string>
+   *   Matching node IDs
+   */
+  private function getNodeIds(?string $excluded_uuid = NULL, array $tags = [], array $categories = [], array $branches = []): array {
+    $query = $this->entityTypeManager->getStorage('node')->getQuery();
+
+    $query->accessCheck(TRUE);
+
+    $query
+      ->condition('type', $this->nodeBundles, 'IN')
+      ->condition('uuid', $excluded_uuid, '<>')
+      ->sort($this->nodeSortField, 'DESC')
+      // We know that we will never need more than the maximum items,
+      // so we will limit the query to this.
+      ->range(0, $this->maxItems);
+
+    if (!empty($tags) || !empty($categories) || empty($branches)) {
+      // Use a condition group for OR logic.
+      $or_group = $query->orConditionGroup();
+
+      // To avoid errors, related to the OR GROUP only containing one condition,
+      // we'll add a fake condition to fill out a possible empty space.
+      $or_group->condition('title', 'ALWAYS_FALSE');
+
+      if (!empty($tags)) {
+        $or_group->condition('field_tags', $tags, 'IN');
+      }
+
+      if (!empty($categories)) {
+        $or_group->condition('field_categories', $categories, 'IN');
+      }
+
+      if (!empty($branches)) {
+        $or_group->condition('field_branch', $branches, 'IN');
+      }
+
+      $query->condition($or_group);
+    }
+
+    return $query->execute();
+  }
+
+  /**
+   * Get matching eventinstance IDs.
+   *
+   * Allows for passing along various term IDs, that we look for in an OR group.
+   * Notice - we only look for the term values on series level, ignoring
+   * inheritance overriding in favor of a simpler codebase.
+   *
+   * @param string|null $excluded_uuid
+   *   A possible entity UUID, that will not get included in results.
+   * @param array<int> $tags
+   *   Tag term IDs, to look for.
+   * @param array<int> $categories
+   *   Category term IDs, to look for.
+   * @param array<int> $branches
+   *   Branch term IDs, to look for.
+   *
+   * @return array<int|string>
+   *   Matching eventinstance IDs
+   *
+   * @throws \Exception
+   */
+  private function getEventInstanceIds(?string $excluded_uuid = NULL, array $tags = [], array $categories = [], array $branches = []): array {
+    $date = new DrupalDateTime('today');
+    $date->setTimezone(new \DateTimezone(DateTimeItemInterface::STORAGE_TIMEZONE));
+    $formatted_date = $date->format(DateTimeItemInterface::DATETIME_STORAGE_FORMAT);
+
+    // Ideally, we'd use EntityQueries instead of a direct DB connection, but
+    // because we need to do some pretty complicated JOINs and subqueries, it
+    // is not an option.
+    $connection = Database::getConnection();
```

Comment:
Please inject the connection in the constructor.

It is available as a service.